acnportal/acnsim/interface.py

```diff
@@ -4,6 +4,154 @@
 import numpy as np
 from datetime import timedelta
 from collections import namedtuple
+from warnings import warn
+
+
+class SessionInfo:
+    """ Class to store information relevant to a charging session.
+
+        Args:
+            station_id (str): Unique identifier of the station (EVSE) where the
+                session takes place.
+            session_id (str): Unique identifier of the charging session.
+            energy_requested (float): Energy requested by the user during the
+                session. [kWh]
+            energy_delivered (float): Energy delivered already during the
+                session. [kWh]
+            arrival (int): Time index when the session begins.
+            departure (int): Time index when the session ends.
+            current_time (int): Time index of the current time.
+            min_rates (Union[float, List[float]): Lower bound for the charging
+                rate of the session. If List (or np.array) length should be
+                departure - arrival and each entry is a lower bound for the
+                corresponding time period.
+            max_rates (Union[float, List[float]): Upper bound for the charging
+                rate of the session. If List (or np.array) length should be
+                departure - arrival and each entry is a upper bound for the
+                corresponding time period.
+    """
+
+    def __init__(
+        self,
+        station_id,
+        session_id,
+        requested_energy,
+        energy_delivered,
+        arrival,
+        departure,
+        estimated_departure=None,
+        current_time=0,
+        min_rates=0,
+        max_rates=float("inf"),
+    ):
+        self.station_id = station_id
+        self.session_id = session_id
+        self.requested_energy = requested_energy
+        self.energy_delivered = energy_delivered
+        self.arrival = arrival
+        self.departure = departure
+        self.estimated_departure = estimated_departure
+        self.current_time = current_time
+        self.min_rates = (
+            np.array([min_rates] * self.remaining_time)
+            if np.isscalar(min_rates)
+            else np.array(min_rates)
+        )
+        self.max_rates = (
+            np.array([max_rates] * self.remaining_time)
+            if np.isscalar(max_rates)
+            else np.array(max_rates)
+        )
+
+    @property
+    def remaining_demand(self):
+        return self.requested_energy - self.energy_delivered
+
+    @property
+    def arrival_offset(self):
+        return max(self.arrival - self.current_time, 0)
+
+    @property
+    def remaining_time(self):
+        offset = max(self.arrival_offset, self.current_time)
```

Comment:
The purpose of arrival_offset is to support the case where EVs have a reservation or are known to arrive sometime in the future. In this case the remaining_time is just the time between the EV will arrive and the departure time. This might be more clear, then we can remove arrival_offset.

```
remaining = min(self.departure - self.arrival, self.departure - self.current_time)
return max(remaining, 0)
```